Libraries/PeriphDrivers/Source/I3C/i3c_reva.c

```diff
@@ -0,0 +1,941 @@
+        }
+
+        /* 3. Optionally write IBIRULES reg to optimize response to incoming IBIs */
+
+        /* 4. Enable controller mode */
+        SET_FIELD(cont_ctrl0, CONT_CTRL0_EN, MXC_V_I3C_REVA_CONT_CTRL0_EN_ON);
+    } else {
+        /* Target mode initialization */
+        /* 1. After reset, write setup registers needed for optional features */
+
+        if (i3c->targ_cap1 & MXC_F_I3C_REVA_TARG_CAP1_RXFIFO_CFG) {
+            MXC_I3C_RevA_SetRXTXThreshold(i3c, MXC_I3C_REVA_RX_TH_NOT_EMPTY,
+                                          MXC_I3C_REVA_TX_TH_ALMOST_FULL);
+        }
+
+        /* Set BCR, DCR */
+        val = 0 << 6; /* Device role: target */
+        if (i3c->targ_cap1 & MXC_S_I3C_REVA_TARG_CAP1_CCCH_LIMITS) {
+            val |= 1 << 0; /* Max data speed limitation */
+        }
+        if (i3c->targ_cap1 & MXC_S_I3C_REVA_TARG_CAP1_CCCH_BASIC) {
+            val |= 1 << 5; /* Supports advanced capabiliries */
+        }
+        if (i3c->targ_cap1 & MXC_S_I3C_REVA_TARG_CAP1_IBI_EVENTS_IBI) {
+            val |= 1 << 1; /* Supports IBI generation */
+        }
+        if (i3c->targ_cap1 & MXC_S_I3C_REVA_TARG_CAP1_IBI_EVENTS_PAYLOAD) {
+            val |= 1 << 2; /* MDB and additional data bytes may follow the IBI */
+        }
+        val |= 1 << 3; /* Offline capable */
+        SET_FIELD(targ_idext, TARG_IDEXT_BUSCHAR, val);
+
+        /* Set DCR to generic device */
+        SET_FIELD(targ_idext, TARG_IDEXT_DEVCHAR, 0);
+
+        /* Enable CCC handling */
+        MXC_I3C_RevA_TargetEnableInt(i3c, MXC_F_I3C_REVA_TARG_INTEN_CCC);
+
+        /* 2. Write 1 to CONFIG.TGTENA */
+        SET_FIELD(targ_ctrl0, TARG_CTRL0_EN, 1);
+
+        /* Clear STOP bit */
+        SET_FIELD(targ_status, TARG_STATUS_STOP, 0);
+    }
+
+    controller[idx].regs = i3c;
+
+    return E_NO_ERROR;
+}
+
+static inline uint8_t MXC_I3C_RevA_GetState(mxc_i3c_reva_regs_t *i3c)
+{
+    return (uint8_t)GET_FIELD(cont_status, CONT_STATUS_STATE);
+}
+
+static inline int MXC_I3C_RevA_WaitForDone(mxc_i3c_reva_regs_t *i3c, int timeout)
+{
+    while (!(i3c->cont_status & MXC_F_I3C_REVA_CONT_STATUS_DONE)) {
+        if (timeout == 0) {
+            return E_TIME_OUT;
+        }
+        if (timeout > 0) {
+            timeout--;
+        }
+    }
+
+    return E_SUCCESS;
+}
+
+int MXC_I3C_RevA_Recover(mxc_i3c_reva_regs_t *i3c)
+{
+    int retries = 3;
+
+    while (retries && MXC_I3C_RevA_GetState(i3c) == MXC_V_I3C_REVA_CONT_STATUS_STATE_SDR_NORM) {
+        MXC_I3C_RevA_EmitStop(i3c);
+        retries--;
+    }
+
+    retries = 20;
+    while (retries-- && GET_FIELD(cont_status, CONT_STATUS_TARG_START)) {
+        MXC_I3C_RevA_AutoIBI(i3c);
+        MXC_I3C_RevA_WaitForDone(i3c, 1000);
+        MXC_I3C_RevA_ClearRXFIFO(i3c);
+    }
+
+    retries = 10;
+    while (retries && MXC_I3C_RevA_GetState(i3c) != MXC_V_I3C_REVA_CONT_STATUS_STATE_IDLE) {
+        retries--;
+    }
+
+    if (retries == 0) {
+        return E_TIME_OUT;
+    }
+
+    return 0;
+}
```

Comment:
Just curious, are the number of retries given by the spec or something we chose that is probably a good amount?

Reply:
For the first one, I observed that it gets stuck there sometimes, i.e., TARG_START is continuously being set even after IBI request is handled. Therefore, I made it fall through after 20 retries, expecting less than 20 devices on the bus. I3C specs do not put a limit on the number of supported devices on an I3C bus. In earlier specs it was given as 11.

Second one is used for delay though I realize now that 10 loops is too short for a delay. The reason I did not want to use MXC_Delay function for delays is that it could conflict with Zephyr OS which also uses SysTick for timing and delays.

Reply:
On a second look, it seems I forgot to clear TARG_START bit. I will try to see if it solves the problem I mentioned.